src/crystal/system/unix/file.cr

```diff
@@ -105,6 +105,25 @@ module Crystal::System::File
     ret
   end
 
+  def self.readlink(path) : String
+    buf = Bytes.new 256
```

Comment:
Reading the [readlink(2)](http://man7.org/linux/man-pages/man2/readlink.2.html) manpage, it asks `lstat` for the actual buffer size and falls back to `PATH_MAX if it's 0. That leads to a race condition between `lstat` and `readlink` and the result can still be truncated, and must be dealt with (retry or raise if buf.size == PATH_MAX).

Reply:
The choice of initial buffer size and multiplication factor are [space, time, race condition] trade offs.

The longest link on my Mac is 138.  On Linux 135.

All requests on a normal system are handled by a single system call with buffer size 256. (Race condition free)
The 2nd request at 1024 will handle what I assume are outliers.
The 3rd requests hits the max on either Linux or Mac at 4096.
The 3 iter limit is designed for systems like GNU Hurd which may have extremely high or no practical limits.

Alternatively I could:
* lstat
* alloc + readlink
* loop on failure?

But this would add a race condition on every call.
Along with the overhead of 2 system calls.
And more heap allocations for the File::Info and underlying stat structure.

Race condition avoidance alone was enough for me to avoid the lstat method.